File: framec_tests/python/src/state_vars/state_vars.py
#Emitted from framec_v0.11.2



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class StateVars:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = StateVarsCompartment('__statevars_state_Init', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        self.return_stack = [None]
        
        # Initialize domain
        
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def X(self,):
        self.return_stack.append(None)
        __e = FrameEvent("X",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def Y(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Y",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def Z(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Z",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $Init
    
    def __statevars_state_Init(self, __e, compartment):
        if __e._message == ">":
            next_compartment = None
            next_compartment = StateVarsCompartment('__statevars_state_A', next_compartment)
            next_compartment.state_vars["x"] = 0
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $A
    
    def __statevars_state_A(self, __e, compartment):
        if __e._message == "X":
            (compartment.state_vars["x"]) = compartment.state_vars["x"] + 1
            return
        elif __e._message == "Y":
            next_compartment = None
            next_compartment = StateVarsCompartment('__statevars_state_B', next_compartment)
            next_compartment.state_vars["y"] = 10
            next_compartment.state_vars["z"] = 100
            self.__transition(next_compartment)
            return
        elif __e._message == "Z":
            next_compartment = None
            next_compartment = StateVarsCompartment('__statevars_state_B', next_compartment)
            next_compartment.state_vars["y"] = 10
            next_compartment.state_vars["z"] = 100
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $B
    
    def __statevars_state_B(self, __e, compartment):
        if __e._message == "X":
            next_compartment = None
            next_compartment = StateVarsCompartment('__statevars_state_A', next_compartment)
            next_compartment.state_vars["x"] = 0
            self.__transition(next_compartment)
            return
        elif __e._message == "Y":
            (compartment.state_vars["y"]) = compartment.state_vars["y"] + 1
            return
        elif __e._message == "Z":
            (compartment.state_vars["z"]) = compartment.state_vars["z"] + 1
            return
    
    # ===================== Actions Block =================== #
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__statevars_state_Init':
            self.__statevars_state_Init(__e, self.__compartment)
        elif self.__compartment.state == '__statevars_state_A':
            self.__statevars_state_A(__e, self.__compartment)
        elif self.__compartment.state == '__statevars_state_B':
            self.__statevars_state_B(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        
    def compartment_info(self):
        return self.__compartment
        

# ===================== Compartment =================== #

class StateVarsCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    

File: framec_tests_old/javascript/src/state_vars/state_vars.js
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class StateVars {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sInit_;
        this.#compartment = new StateVarsCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    X() {
        let e = FrameEvent("X",null);
        this.#mux(e);
    }
    
    Y() {
        let e = FrameEvent("Y",null);
        this.#mux(e);
    }
    
    Z() {
        let e = FrameEvent("Z",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sInit_:
                this.#sInit_(e);
                break;
            case this.#sA_:
                this.#sA_(e);
                break;
            case this.#sB_:
                this.#sB_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sInit_(e) {
        switch (e._message) {
            case ">":
                {
                let compartment =  new StateVarsCompartment(this.#sA_);
                
                compartment.StateVars["x"] = this.#compartment.StateVars["x"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sA_(e) {
        switch (e._message) {
            case "X":
                {
                (this.#compartment.StateVars["x"]) = (this.#compartment.StateVars["x"]) + 1;
                
                return;
                }
                
            case "Y":
                {
                let compartment =  new StateVarsCompartment(this.#sB_);
                
                compartment.StateVars["y"] = this.#compartment.StateVars["y"] + 1;
                compartment.StateVars["z"] = this.#compartment.StateVars["z"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Z":
                {
                let compartment =  new StateVarsCompartment(this.#sB_);
                
                compartment.StateVars["y"] = this.#compartment.StateVars["y"] + 1;
                compartment.StateVars["z"] = this.#compartment.StateVars["z"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sB_(e) {
        switch (e._message) {
            case "X":
                {
                let compartment =  new StateVarsCompartment(this.#sA_);
                
                compartment.StateVars["x"] = this.#compartment.StateVars["x"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Y":
                {
                (this.#compartment.StateVars["y"]) = (this.#compartment.StateVars["y"]) + 1;
                
                return;
                }
                
            case "Z":
                {
                (this.#compartment.StateVars["z"]) = (this.#compartment.StateVars["z"]) + 1;
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    compartment_info() {
        return this.#compartment
    }
    
    
};

//=============== Compartment ==============//

class StateVarsCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class StateVarsController extends StateVars {

	constructor() {
	  super()
	}
};

********************/

module.exports = StateVars
